[Willy Angga] iP

willyamped opened the pull request:
# Amped
> “Your mind is for having ideas, not holding them.” – David Allen [(source)](https://dansilvestre.com/productivity-quotes)

Amped is
- text-based
- easy to learn
- ~~FAST~~ *SUPER FAST* to use

All you need to do is,

1. download the JAR file.
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :grin:

And it's **FREE!**

Features:

 - [x] Managing tasks
 - [ ] Managing deadlines (coming soon)
 - [ ] Reminders (coming soon)
 
If you Java programmer, you can use it to practice Java too. Here's the `main` method:
```java
public static void main(String[] args) {
        Duke duke = new Duke("./data/duke.txt");
        duke.run();
}
```

Commit:
Add Gradle support

willyamped committed:
Level-1

willyamped committed:
Level-2

willyamped committed:
added Todo, Event, Deadline

willyamped committed:
Revert "added Todo, Event, Deadline"

This reverts commit a597d5f2d36f03e516279409c4bc163a2f63ad5f.

willyamped committed:
Undo the revert

willyamped committed:
Added 3 exceptions

willyamped committed:
added comprehensive welcome message

willyamped committed:
add delete and fix exception bugs

willyamped committed:
Add documentation

willyamped committed:
Add automated-testing, bug fixes

willyamped committed:
Add isNumeric method, Fix todo

willyamped committed:
Add readFile(), saveFile() method

willyamped committed:
Fix getDeadline to recognise date and time

willyamped committed:
Merge branch 'branch-Level-7'

willyamped committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java

willyamped committed:
Fix readFile()

willyamped committed:
Add taskList

willyamped committed:
Add storage

willyamped committed:
Move getDeadline() and getDescription() to TaskList

willyamped committed:
Add parser

willyamped committed:
Add Ui

willyamped committed:
Add package

willyamped committed:
Add JUnit tests

willyamped committed:
Add testGetSize() to TaskListTest

willyamped committed:
Add findTask()

willyamped committed:
Add JAR File

willyamped committed:
Commit MANIFEST.MF

willyamped committed:
Add JavaDoc

willyamped committed:
Fix coding style

willyamped committed:
Add find() method

willyamped committed:
Merge branch 'branch-A-JavaDoc'

willyamped committed:
Merge branch 'branch-A-CodingStandard'

willyamped committed:
Merge branch 'branch-Level-9'

willyamped committed:
Improve javadocs convention

willyamped committed:
Fix javadocs and parse

willyamped committed:
Fix findTask()

willyamped committed:
Fix findTask() to allow a single word

willyamped committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d'

tsinyee commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+                    throw new InvalidCommandException("Command not Found");
+                }
+                storage.saveFile(taskList);
+            } catch (InvalidCommandException e) {
+                System.out.println(e.toString());
+            } catch (EmptyDescriptionException e) {
+                System.out.println(e.toString());
+            } catch (InvalidValueException e) {
+                System.out.println(e.toString());
+            } catch (DateTimeParseException e) {
+                System.out.println(e.getMessage());
+            }
```
consider minimizing Code Duplication here. perhaps you could create a parent DukeException class that all the other exception classes inherit from.
```suggestion
            } catch (DukeException e) {
                System.out.println(e.getMessage());
            } 
```

tsinyee replied:
code duplication. can consider creating a parent class DukeException which all the other exception classes inherit from 
```suggestion
            } catch (DukeException e) {
                System.out.println(e.toString());
            } 
```

tsinyee commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+     */
+    public void parseDelete(String[] arr) throws InvalidCommandException,
+            InvalidValueException {
+        if (arr.length == 1) {
+            throw new InvalidCommandException("Please specify a number");
+        } else if (!isNumeric(arr[1])) {
+            throw new InvalidCommandException("Please enter a number");
+        } else if (taskList.getSize() == 0) {
+            throw new InvalidCommandException("You have not added any task!");
+        } else if ((Integer.parseInt(arr[1]) > taskList.getSize()
+                || Integer.parseInt(arr[1]) <= 0)) {
+            throw new InvalidValueException("Enter a valid number!");
```
code here is duplicated from the `parseDone(String[] arr)` method.  you can consider creating  a
method to extract and validate the index

tsinyee replied:
code here is similar to `parseDone`. can consider creating a method to extract and validate the index

tsinyee commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,117 @@
+     *
+     * @param ls A list to keep track of all tasks.
+     */
+    public void readFile(TaskList ls) {
```
method is abit too long and has quite a number of nested statements. you can create another method to convert the text format in storage to tasks.

tsinyee replied:
method is too long and has quite a number of nested statements. an consider creating a method to convert the text format in storage to tasks.

tsinyee commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,203 @@
+            try {
+                input = scan.nextLine();
+                arr = input.split(" ");
+                if (input.equals("bye")) {
+                    break;
+                } else if (arr[0].equals("done")) {
+                    parseDone(arr);
+                } else if (arr[0].equals("delete")) {
+                    parseDelete(arr);
+                } else if (arr[0].equals("todo")) {
+                    parseTodo(arr);
+                } else if (arr[0].equals("deadline")) {
+                    parseDeadline(arr);
+                } else if (arr[0].equals("event")) {
+                    parseEvent(arr);
+                } else if (input.equals("list")) {
+                    parseList();
+                } else if (arr[0].equals("find")) {
+                    parseFind(arr);
```
what if user enters "List"? can consider using the method `equalsIgnoreCase()`

tsinyee reviewed (commented):
I like that your method names are very clear and follow the coding standards! 👍   
However some methods are too long and can be shortened. LGTM, almost. Just a few nits to fix.

willyamped committed:
Add GUI

willyamped committed:
Fix GUI slightly

willyamped committed:
Merge branch 'branch-Level-10'

willyamped committed:
Fix GUI and add javadocs

willyamped committed:
Fix delete, bye

willyamped committed:
Change welcome message and images

willyamped committed:
Add JAR with gradle

willyamped committed:
Add assertions

WuaaAj commented on src/main/java/duke/Task.java:
```diff
@@ -0,0 +1,73 @@
+public class Task {
+    protected String description;
+    protected boolean isDone;
+    protected String deadline = "";
```
I think should be better to move variable "deadline" to "Deadline" class. Since not all kinds of tasks have such property.

WuaaAj commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,30 @@
+ * Represents a task that has a description and deadline.
+ */
+public class Deadline extends Task {
+    protected String by;
```
Can have a better name for "by". Maybe "deadline"?

WuaaAj commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,30 @@
+ * Represents a task that has a description and time.
+ */
+public class Event extends Task {
+    protected String by;
```
Can consider changing the name.

WuaaAj commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,204 @@
+                    return parseList();
+                } else if (arr[0].equals("find")) {
+                    return parseFind(arr);
+                } else {
```
I think the "else" condition should also through an "InvalidCommandException".

WuaaAj reviewed (commented):
I find your code is very readable and neat! 
Just notice a few things that can be improved.

luffingluffy commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,204 @@
+    }
+
+    /**
+     * Makes sense of the user input.
```
I think this can be changed into something more active; ie: converts the user's inputs.

luffingluffy commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,117 @@
+                arr = data.split(" ");
+                String task = arr[0];
+                String description = arr[1] + " ";
+                if (task.charAt(2) == 'D') {
```
I think you can use a switch statement instead.

luffingluffy commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,192 @@
+        String str = "";
+        boolean bool = false;
+        for(int i = 1; i < arr.length; i++) {
+            if (!bool) {
```
Similarly, I think you can use a switch statement here.

luffingluffy reviewed (commented):
I think you can change some if-else statements to switch statements. Otherwise, LGTM!

willyamped committed:
Add more assertions and fix javadocs header

willyamped committed:
Improve code quality

willyamped committed:
Merge pull request #1 from willyamped/branch-A-CodeQuality

Improve code quality

willyamped committed:
Merge branch 'master' of https://github.com/willyamped/ip

willyamped committed:
Merge branch 'master' into branch-A-Assertions

willyamped committed:
Merge pull request #2 from willyamped/branch-A-Assertions

Branch a assertions

willyamped committed:
Extract a method check() for parseDone and parseDelete

The code is duplicated and hence extracting the method is a good idea

willyamped committed:
Fix code based on peer review suggestions

Fix code in Parser.java, Event.java, Deadline.java

willyamped committed:
Remove unused variables

willyamped committed:
Fix readFile()

Previously, when there are more than 1 word for the description, the readFile() did not work properly.
After fixing the readFile() by using a counter, it can now read any task with any description correctly.

willyamped committed:
Add search feature

Previously, the user can only find a task that exactly matches the keyword. For example, typing "swim" when there is a "swimming" task will return an empty list.
This current feature allows a user to find a task that exactly/ partially matches a search term.

willyamped committed:
Add Ui.png

willyamped committed:
Change Ui.png

willyamped committed:
Change Ui.png again

willyamped committed:
Remove code from Duke.java

willyamped committed:
Update README.md

willyamped committed:
Fix delete() to add a message on the number of tasks

willyamped committed:
Improve code quality by extracting methods in Storage.java

willyamped committed:
Make further changes to readFile() in Storage.java

willyamped committed:
Change filepath

willyamped committed:
Update README.md

willyamped committed:
Update README.md

willyamped committed:
Update README.md

willyamped committed:
Set theme jekyll-theme-cayman

willyamped committed:
Fix JUnit and add checkstyle

willyamped committed:
Merge branch 'master' of https://github.com/willyamped/ip

willyamped committed:
Update README.md

willyamped committed:
Move methods from TaskList to Ui

willyamped committed:
Merge branch 'master' of https://github.com/willyamped/ip
